src/gui/editorwidgets/qgsvaluerelationconfigdlg.cpp

```diff
@@ -70,6 +71,23 @@ QVariantMap QgsValueRelationConfigDlg::config()
 void QgsValueRelationConfigDlg::setConfig( const QVariantMap &config )
 {
   QgsVectorLayer *lyr = QgsProject::instance()->mapLayer<QgsVectorLayer *>( config.value( QStringLiteral( "Layer" ) ).toString() );
+  // If layer is null, let's try to match it against name and provider
+  if ( ! lyr )
+  {
+    const auto name { config.value( QStringLiteral( "LayerName" ) ).toString() };
+    if ( ! name.isEmpty() )
+    {
+      for ( const auto &l : QgsProject::instance()->mapLayers( true ) )
```

Comment:
1. take a const copy
2. Make type explicit

Reply:
>     2. Make type explicit

I'd ask for a bit more of flexibility, there isn't any danger using `auto` here.

Reply:
But there is also no advantage to use it and having it explicit avoids potential danger and makes it easier to review and read the code.

I'm already starting to ask myself questions (like: how is that ref after converted to a pointer on the line below? Is that a QMap<> something or a QgsMapLayer in that `l` here? Where I only found the QMap possibility after looking it up in the source ... ). I think that would be easier if it was explicit ;)

Reply:
> But there is also no advantage to use it and having it explicit avoids potential danger and makes it easier to review and read the code.

There is actually an advantage for who **writes** the code: that she can just write `const auto &something` and forget about what it is. 

But I can see the advantage for who **reads** the code.

Reply:
So given that code is written once and read many times I'd be in favour of thinking of the reader at writing time :)